openhands/core/main.py

```diff
@@ -96,7 +96,7 @@ async def run_controller(
 
     Args:
         config: The app config.
-        task_str: The task to run. It can be a string.
+        initial_user_actions: a list of Actions containing initial user inputs
```

Comment:
Just wondering, a single MessageAction's `content` is a list of both strings and `image_urls`, so we can implement accepting images by sending a MessageAction, not necessarily a list of MessageActions, or am I missing something?

Reply:
A single MessageAction object contains a single string (content) and a list of image_urls. So I think having a single MessageAction object should also work if we are sure that we always have only a single string in input. But I think it would be best to ask @xingyaoww  because I am relatively new to this code base.

Reply:
Ah, you're right, @adityasoni9998. This is a bit confusing though. Quite a number of places in the code assume that the "task" is a single string. A MessageAction can only be created with a single TextContent and multiple image_urls, and CodeAct code uses it clearly as single [here](https://github.com/All-Hands-AI/OpenHands/blob/f994277d0f17214693e7657291705dbcae944e5a/agenthub/codeact_agent/codeact_agent.py#L119). But we _can_ add another TextContent to any user message, and [do it here](https://github.com/All-Hands-AI/OpenHands/blob/f994277d0f17214693e7657291705dbcae944e5a/agenthub/codeact_agent/codeact_agent.py#L283), so multiple TextContents... are kind of supported - in the message infrastructure. 🤔 

So I guess my question becomes, what are the consequences if we try to work with a list of consecutive MessageActions as "task"? The code assumes a single "task" and a single event (MessageAction) in agents. Some code looks for a single string [from a dict](https://github.com/All-Hands-AI/OpenHands/blob/42be4ee5bc26afcd40f2dd5dda8d1536e9e61b4a/agenthub/browsing_agent/browsing_agent.py#L207) too.

At first sight, we may want to consider two other details:

- There are LLMs that don't support two or more actions with the same `role` consecutively. So CodeAct [merges the adjacent actions](https://github.com/All-Hands-AI/OpenHands/blob/f994277d0f17214693e7657291705dbcae944e5a/agenthub/codeact_agent/codeact_agent.py#L252) in one. So the MessageActions in the list will - and have to - end up merged either way, when they get sent to the LLM.

- [this](https://github.com/All-Hands-AI/OpenHands/blob/f994277d0f17214693e7657291705dbcae944e5a/openhands/controller/state/state.py#L160) is "the task" as used in some places, including BrowsingAgent. I don't know, that might need some tweaking if our task becomes a list. 🤔

Reply:
I think using one MessageAction as task input is actually a great idea 🤔 - maybe we can go with that?

Reply:
Ok, I have changed the code to have a single MessageAction object as the task input.